File: ios/ReactNativeModules/IntercomModule.swift
//
//  IntercomModule.swift
//  connectreactive
//
//  Created by Sergei Golishnikov on 03/06/2021.
//

import Foundation
import React
import Intercom

@objc(IntercomModule)
class IntercomModule: RCTEventEmitter {
    
    override func supportedEvents() -> [String]! { [] }
    override class func requiresMainQueueSetup() -> Bool {
        return false
    }
    
    @objc
    func registerUnidentifiedUser() {
        Intercom.registerUnidentifiedUser()
    }
    
    @objc
    func setLauncherVisibility(_ isVisible: Bool) {
        Intercom.setLauncherVisible(isVisible)
    }
    
    @objc
    func loginUser(_ userId: String) {
        Intercom.registerUser(withUserId: userId)
    }
    
    @objc
    func logoutUser() {
        Intercom.logout()
    }
    
    @objc
    func presentIntercom() {
        Intercom.presentMessenger()
    }
    
    @objc
    func presentIntercomCarousel() {
        Intercom.presentCarousel("19961407")
    }
}

File: android/app/src/main/java/com/connect/club/react/bridge/iosLike/IntercomModule.kt
package com.connect.club.react.bridge.iosLike

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import io.intercom.android.sdk.Intercom
import io.intercom.android.sdk.identity.Registration

class IntercomModule(reactContext: ReactApplicationContext?) :
    ReactContextBaseJavaModule(reactContext) {
    override fun getName() = "IntercomModule"

    @ReactMethod
    fun registerUnidentifiedUser() {
        debugP("Intercom register unidentified")
        Intercom.client().registerUnidentifiedUser()
    }

    @ReactMethod
    fun setLauncherVisibility(isVisible: Boolean?) {
        debugP("Intercom set launcher visible:", isVisible)
        Intercom.client().setLauncherVisibility(
            if (isVisible == true) Intercom.Visibility.VISIBLE else Intercom.Visibility.GONE
        )
    }

    @ReactMethod
    fun loginUser(userId: String) {
        debugP("Intercom login:", userId)
        val registration = Registration.create().withUserId(userId)
        Intercom.client().registerIdentifiedUser(registration)
    }

    @ReactMethod
    fun logoutUser() {
        debugP("Intercom logout")
        Intercom.client().logout()
    }

    @ReactMethod
    fun presentIntercom() {
        debugP("Intercom present")
        Intercom.client().displayMessenger()
    }

    @ReactMethod
    fun presentIntercomCarousel() {
        debugP("Intercom present carousel")
        Intercom.client().displayCarousel("19961407")
    }
}